design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,163 @@
+  running: true
+  template:
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+kubelet will emit following event:
+```
+SMT Alignment Error: requested 9 cpus not multiple cpus per core = 2
+```
+The following picture helps show how this is a problem
+<img src="smt-slignment-error-diagram.png">
+
+## Goals
+Fix [BZ#2228103](https://bugzilla.redhat.com/show_bug.cgi?id=2228103) on the Kubevirt scope.
+
+## Non  Goals
+Fix the actual scheduling of the virt-launcher to a node that will reject it.
+
+## Definition of Users
+This bug addresses all KubeVirt users running VMs with isolateEmulatorThread on nodes with hyperthreading enabled - specifically DPDK and Realtime users.
+
+## User Stories
+A user should be able to run VM with dedicatedCPUPlancement and isolateEmulatorThread enabled on a node with hyperthreading enabled, regardless of the amount of guest CPUs requested.
+
+## Repos
+- [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Bug fix design proposal
+The bug fix proposal introduces an alpha annotation `kubevirt.io/CPUManagerPolicyBetaOptions:full-pcpus-only`:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    metadata:
+      annotations:
+        kubevirt.io/CPUManagerPolicyBetaOptions:full-pcpus-only
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+
+When the VM is configured with `isolateEmulatorThread: true` and the above annotation, the virt-controller will set the virt-launcher pod with an even amount of CPUs, meaning:
+
+| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) |
+|-----------------------------------|------------------|-----------------------|
+| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         |
+
+Notes:
+1. In both scenarios the guest will get the same amount of CPUs it requested.
+2. For simplicity reasons, the housekeeping cgroup will contain only 1 CPU. However, this is open for discussion.
+
+Pros:
+* Cleaner user experience. No prior knowledge of CPUs added behind the scene is needed. 
+* Matches the current crio convention of adding annotations in order to get high-performance pods, like `cpu-load-balancing.crio.io: disable`
+* Reduces the “wasted” CPUs to minimum possible per scenario, i.e. only for VMs that specifically add this annotation, and only when they ask for even CPUs. 
+* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.
+
+Cons:
+* We’re adding yet another semi alpha knob to an already complicated set of knobs.
+
+# Fix Alternatives
+## Alternative #1 Adding `full-pcpus-only` option as a global configuration
```

Comment:
@RamLavi I am not sure whether we can consider it as alternative, because it depends on the main proposal. It can rather be an option in the implementation details, to enhance the kubevirt CR with the option to annotate every VM with the suggested annotation.

Reply:
>@RamLavi I am not sure whether we can consider it as alternative, because it depends on the main proposal. It can rather be an option in the implementation details, to enhance the kubevirt CR with the option to annotate every VM with the suggested annotation.

I don't mean that this global option will annotate every VM once it is enabled. 
I mean that this will be a global behavior, in a similar way that `defaultRuntimeClass` sets the runtimeClass without changing the VM manifest (metadata or spec API) at all.
It is in that sense - that it is a different alternative.